prober/icmp.go

```diff
@@ -62,23 +62,44 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 	if ip.IP.To4() == nil {
 		requestType = ipv6.ICMPTypeEchoRequest
 		replyType = ipv6.ICMPTypeEchoReply
+
 		socket, err = icmp.ListenPacket("ip6:ipv6-icmp", "::")
+		if err != nil {
+			level.Error(logger).Log("msg", "Error listening to socket", "err", err)
+			return
+		}
 	} else {
 		requestType = ipv4.ICMPTypeEcho
 		replyType = ipv4.ICMPTypeEchoReply
-		socket, err = icmp.ListenPacket("ip4:icmp", "0.0.0.0")
-	}
 
-	if err != nil {
-		level.Error(logger).Log("msg", "Error listening to socket", "err", err)
-		return
+		if !module.ICMP.DontFragment {
+			socket, err = icmp.ListenPacket("ip4:icmp", "0.0.0.0")
+			if err != nil {
+				level.Error(logger).Log("msg", "Error listening to socket", "err", err)
+				return
+			}
+		} else {
+			s, err := net.ListenPacket("ip4:icmp", "0.0.0.0")
+			if err != nil {
+				level.Error(logger).Log("msg", "Error listening to socket", "err", err)
+				return
+			}
+
+			rc, err := ipv4.NewRawConn(s)
+			if err != nil {
+				level.Error(logger).Log("msg", "cannot construct raw connection", "err", err)
+				return
+			}
+			socket = &dfConn{c: rc}
+		}
 	}
+
 	defer socket.Close()
 
 	var data []byte
-	if payload != 0 {
-		data = make([]byte, payload)
-		copy(data[:], "Prometheus Blackbox Exporter")
+	if module.ICMP.PayloadSize != 0 {
+		data = make([]byte, module.ICMP.PayloadSize)
+		copy(data, "Prometheus Blackbox Exporter")
```

Comment:
What if the payload size is smaller than this string?

Reply:
https://play.golang.org/p/qrDFrlW6He Then only the payload size will be sen't with truncated string.